From aefe6b917b6b11dca8ff552ba56eca3e42e713ea Mon Sep 17 00:00:00 2001
From: Lim Siew Hoon <siew.hoon.lim@intel.com>
Date: Thu, 3 Oct 2019 01:48:43 +0800
Subject: [PATCH 2/2] msdkdec: output the decoded frame immediately if decoded 
 order is required

DecodedOrder was deprecated in msdk-2017 version, but some customers
still use this for low-latency streaming of non-b-frame encoded streams,
which needs to output the frame at once

Port from master branch to 1.16.0 code based.

Signed-off-by: Lim Siew Hoon <siew.hoon.lim@intel.com>
---
 sys/msdk/gstmsdkdec.c | 17 ++++++++++++++++-
 1 file changed, 16 insertions(+), 1 deletion(-)

diff --git a/sys/msdk/gstmsdkdec.c b/sys/msdk/gstmsdkdec.c
index 0870cbabb..8fad44680 100644
--- a/sys/msdk/gstmsdkdec.c
+++ b/sys/msdk/gstmsdkdec.c
@@ -896,7 +896,14 @@ gst_msdkdec_handle_frame (GstVideoDecoder * decoder, GstVideoCodecFrame * frame)
     bitstream.Data = map_info.data;
     bitstream.DataLength = map_info.size;
     bitstream.MaxLength = map_info.size;
-    bitstream.DataFlag = MFX_BITSTREAM_COMPLETE_FRAME;
+    /*
+     * MFX_BITSTREAM_COMPLETE_FRAME was removed since commit df59db9, however
+     * some customers still use DecodedOrder (deprecated in msdk-2017 version)
+     * for low-latency streaming of non-b-frame encoded streams, which needs to
+     * output the frame at once, so add it back for this case
+     */
+    if (thiz->param.mfx.DecodedOrder == GST_MSDKDEC_OUTPUT_ORDER_DECODE)
+      bitstream.DataFlag |= MFX_BITSTREAM_COMPLETE_FRAME;
   } else {
     /* Non packetized streams: eg: vc1 advanced profile with per buffer bdu */
     gst_adapter_push (thiz->adapter, gst_buffer_ref (input_buffer));
@@ -1069,6 +1076,14 @@ gst_msdkdec_handle_frame (GstVideoDecoder * decoder, GstVideoCodecFrame * frame)
     flow = GST_FLOW_OK;
   }
 
+  /*
+   * DecodedOrder was deprecated in msdk-2017 version, but some
+   * customers still using this for low-latency streaming of non-b-frame
+   * encoded streams, which needs to output the frame at once
+   */
+  if (thiz->param.mfx.DecodedOrder == GST_MSDKDEC_OUTPUT_ORDER_DECODE)
+    gst_msdkdec_finish_task (thiz, task);
+
 done:
   if (surface)
     free_surface (thiz, surface);
-- 
2.13.6

